File: www/server1/cgi-bin/upload_file.php
<?php
// Set the default timezone
date_default_timezone_set('UTC');

// Log file path
$logFile = './debug.log';

// Function to log messages to the debug log file
function logMessage($message) {
    global $logFile;
    $timestamp = date('Y-m-d H:i:s');
    file_put_contents($logFile, "[$timestamp] $message\n", FILE_APPEND);
}

// Function to create a dark mode HTML response
function createHtmlResponse($message, $success = true) {
    $status = $success ? "Success" : "Error";
    return "
    <!DOCTYPE html>
    <html lang=\"en\">
    <head>
        <meta charset=\"UTF-8\">
        <meta name=\"viewport\" content=\"width=device-width, initial-scale=1.0\">
        <title>File Upload $status</title>
        <style>
            body {
                background-color: #121212;
                color: #ffffff;
                font-family: Arial, sans-serif;
                display: flex;
                justify-content: center;
                align-items: center;
                height: 100vh;
                margin: 0;
            }
            .container {
                text-align: center;
                padding: 20px;
                border: 1px solid #333;
                border-radius: 10px;
                background-color: #1e1e1e;
            }
            .container h1 {
                color: #4caf50;
            }
            .container p {
                color: #ffffff;
            }
        </style>
    </head>
    <body>
        <div class=\"container\">
            <h1>$status</h1>
            <p>$message</p>
        </div>
    </body>
    </html>";
}

// Function to read raw POST data from stdin
function getRawPostData() {
    $rawData = '';
    while ($chunk = fread(STDIN, 1024)) {
        $rawData .= $chunk;
    }
    return $rawData;
}

// Function to extract the boundary from the content type header
function getBoundary($contentType) {
    if (preg_match('/boundary=(.*)$/', $contentType, $matches)) {
        return $matches[1];
    }
    return null;
}

function parseMultipartData($rawData, $boundary) {
    $boundary = '--' . $boundary;
    $endBoundary = $boundary . '--';
    $parts = explode($boundary, $rawData);

    foreach ($parts as $part) {
        if (empty(trim($part)) || trim($part) == '--') {
            continue;
        }
        if (preg_match('/Content-Disposition: form-data; name="fileToUpload"; filename="([^"]+)"/', $part, $matches)) {
            $filename = $matches[1];
            $part = preg_replace('/Content-Disposition:.*filename="[^"]+"/', '', $part);
            $part = preg_replace('/Content-Type:.*\s/', '', $part, 1);
            $fileContent = trim($part);
            return [$filename, $fileContent];
        }
    }
    return [null, null];
}

// Function to save the file data to the specified directory
function saveFile($filename, $fileData) {
    $timestamp = time();
    $newFileName = "user_{$timestamp}_{$filename}";
    $destination = './www/server1/images/' . $newFileName;
    if (file_put_contents($destination, $fileData) !== false) {
        logMessage("File uploaded successfully: $newFileName");
        return $newFileName;
    } else {
        logMessage("Failed to save uploaded file.");
        return null;
    }
}

// Main script execution
$contentType = getenv('CONTENT_TYPE');
$rawData = getRawPostData();
logMessage("Raw POST data length: " . strlen($rawData));

$boundary = getBoundary($contentType);
if ($boundary) {
    logMessage("Boundary found: $boundary");
    list($filename, $fileData) = parseMultipartData($rawData, $boundary);

    if ($filename && $fileData) {
        $savedFileName = saveFile($filename, $fileData);
        if ($savedFileName) {
            echo createHtmlResponse("File uploaded successfully: $savedFileName");
            exit;
        } else {
            echo createHtmlResponse("Failed to save uploaded file.", false);
            exit;
        }
    } else {
        logMessage("No valid file part found in the request.");
        echo createHtmlResponse("No valid file part found in the request.", false);
        exit;
    }
} else {
    logMessage("No boundary found in the content type.");
    echo createHtmlResponse("No boundary found in the content type.", false);
    exit;
}
?>

File: www/server1/cgi-bin/upload_file.py
#!/usr/bin/env python3
import os
import sys
import hashlib
import time
from datetime import datetime

def create_html_response(message, success=True):
    """Function to create a dark mode HTML response."""
    status = "Success" if success else "Error"
    return f"""
    <!DOCTYPE html>
    <html lang="en">
    <head>
        <meta charset="UTF-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>File Upload {status}</title>
        <style>
            body {{
                background-color: #121212;
                color: #ffffff;
                font-family: Arial, sans-serif;
                display: flex;
                justify-content: center;
                align-items: center;
                height: 100vh;
                margin: 0;
            }}
            .container {{
                text-align: center;
                padding: 20px;
                border: 1px solid #333;
                border-radius: 10px;
                background-color: #1e1e1e;
            }}
            .container h1 {{
                color: #4caf50;
            }}
            .container p {{
                color: #ffffff;
            }}
        </style>
    </head>
    <body>
        <div class="container">
            <h1>{status}</h1>
            <p>{message}</p>
        </div>
    </body>
    </html>
    """

def get_raw_post_data():
    """Function to read raw POST data from stdin in binary mode until EOF is encountered."""
    raw_data = b''
    while True:
        chunk = sys.stdin.buffer.read(4096)
        if not chunk:
            break
        raw_data += chunk

    return raw_data

def check_file_data(file_data):
	"""Function to check if the file data is empty."""
	if not file_data:
		print(create_html_response("No file data received.", False))
		sys.exit(1)
	"""Check if there is spaces inside the filename."""
	filename = os.getenv('FILENAME')
	if ' ' in filename:
		print(create_html_response("Filename cannot contain spaces.", False))
		sys.exit(1)
	"""Check the filetype if it a forbidden file type."""
	file_type = filename.split('.')[-1]
	forbidden_file_types = ['php', 'py', 'pl', 'sh', 'rb', 'cgi', 'java', 'jsp', 'asp', 'aspx', 'html', 'htm', 'shtml', 'xhtml', 'xml', 'css', 'js', 'json', 'htaccess', 'conf', 'htpasswd', 'log', 'bak', 'sql', 'sqlite', 'db', 'dbf', 'csv', 'ini', 'sh', 'bat', 'cmd', 'ps1', 'psm1', 'psd1', 'ps1xml', 'psc1', 'pssc', 'msh', 'msh1', 'msh2', 'mshxml', 'msh1xml', 'msh2xml', 'scf', 'lnk', 'inf', 'reg', 'dll', 'vbs', 'vbe', 'js', 'jse', 'ws', 'wsf', 'wsc', 'wsh', 'ps', 'psc1', 'psc2', 'msh', 'msh1', 'msh2', 'mshxml', 'msh1xml', 'msh2xml', 'scf', 'lnk', 'inf', 'reg', 'dll', 'vbs', 'vbe', 'js', 'jse', 'ws', 'wsf', 'wsc', 'wsh', 'ps', 'psc1', 'psc2', 'msh', 'msh1', 'msh2', 'mshxml', 'msh1xml', 'msh2xml', 'scf', 'lnk', 'inf', 'reg', 'dll', 'vbs', 'vbe', 'js', 'jse', 'ws', 'wsf', 'wsc', 'wsh', 'ps', 'psc1', 'psc2', 'msh', 'msh1', 'msh2', 'mshxml', 'msh1xml', 'msh2xml', 'scf', 'lnk', 'inf', 'reg', 'dll', 'vbs', 'vbe', 'js', 'jse', 'ws', 'wsf', 'wsc', 'wsh', 'ps', 'psc1', 'psc2']
	for file_type in forbidden_file_types:
		if file_type in filename:
			print(create_html_response("File type not allowed.", False))
			sys.exit(1)

def save_file(filename, file_data):
    """Function to save the file data to the specified directory."""
    new_file_name = f"user_{int(time.time())}_{filename}"
    destination = f'./www/server1/images/{new_file_name}'
    try:
        with open(destination, 'wb') as file:
            file.write(file_data)

        return new_file_name

    except IOError as e:
        return None

# Main script execution
filename = os.getenv('FILENAME')
file_data = get_raw_post_data()

check_file_data(file_data)

if filename:
    saved_file_name = save_file(filename, file_data)
    if saved_file_name:
        print(create_html_response(f"File uploaded successfully: {saved_file_name}"))
    else:
        print(create_html_response("Failed to save uploaded file.", False))
else:
    print(create_html_response("Filename not provided.", False))
